### PR TITLE
Ext new format

### PR DESCRIPTION
**Pull request by yatli**
@jmamma we still need to adapt the editor to use it.

~~but given the current situation we may need to first casecade back to piano roll branch.~~

edit: since there are no new stuff in piano_roll we can just proceed on this branch..

### CODE REVIEW

**Comment by yatli** [comment]
The ext track size is reduced from 2000+ bytes to 1600+ bytes, also gaining the following capabilities:

- 8 lock targets
- up to 16 events per step (lock + note)
- individual utiming for each event
- lock engage/disengage behavior: a lock-engage event is sticky, like conventional automation in a piano roll. Use another lock-disengage event to recall the original lock value (the plock behavior)

The actual number of slots to keep events is still 512 -- we're still able to fill the track with 4x128 midi notes. But now locks and notes share the same event storage.

Wrt the lock engage/disengage behavior: I designed it so that we overcome the inconsistency of long notes vs. single-step p-lock.
The actual behavior is now left to the track editor -- both sticky/plock behavior can be described with this storage model.

**Comment by yatli** [comment]
Afk for now. @jmamma can you take a look at the implementation and see if you can work out SeqExtStepPage?

**Comment by yatli** [comment]
I'll unpack my A4 when I get home...

**Comment by jmamma** [comment]
https://en.wikipedia.org/wiki/Nibbles_(video_game)

First video game I played on the PC.

File: avr/cores/megacommand/CommonTools/NibbleArray.h [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	void shift_right(uint8_t len) {
+		if (len == 0) return;
+		NibbleArray<N> that = *this;
+		for(uint8_t i = 1; i < len; ++i) {
+			set(i, that.get(i-1));
+		}
+		set(0, that.get(len-1));
+	}
+
+	void reverse(uint8_t len) {
```
Comment (jmamma) [inline avr/cores/megacommand/CommonTools/NibbleArray.h:54]: This implementation won't work. Once you get to i/2 you have already overwritten the remaining values.

File: avr/cores/megacommand/CommonTools/NibbleArray.h [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+		for(uint8_t i = 0; i < len; ++i) {
+			set(i, that.get(i+1));
+		}
+		set(len, that.get(0));
+	}
+
+	void shift_right(uint8_t len) {
+		if (len == 0) return;
+		NibbleArray<N> that = *this;
+		for(uint8_t i = 1; i < len; ++i) {
```
Comment (jmamma) [inline avr/cores/megacommand/CommonTools/NibbleArray.h:48]: this loop needs be reversed?

File: avr/cores/megacommand/MCL/ExtSeqTrack.cpp [hunk 3]
```diff
@@ -31,6 +29,63 @@ void ExtSeqTrack::re_sync() {
...
+uint16_t ExtSeqTrack::add_event(uint8_t step) {
+  uint8_t u = timing_buckets.get(step); 
+  if (15 == u || event_count >= NUM_EXT_EVENTS) {
+    // bucket full or track full
+    return 0xFFFF;
+  }
+  ++event_count;
+  timing_buckets.set(step, u+1);
+  uint16_t idx, end;
+  locate(step, idx, end);
```
Comment (jmamma) [inline avr/cores/megacommand/MCL/ExtSeqTrack.cpp:56]: ~idx, end should be a &idx, &end~

File: avr/cores/megacommand/CommonTools/NibbleArray.h (hunk 1 above)
Comment (yatli) [reply avr/cores/megacommand/CommonTools/NibbleArray.h:54]: "that" is copied

File: avr/cores/megacommand/MCL/ExtSeqTrack.cpp (hunk 3 above)
Comment (yatli) [reply avr/cores/megacommand/MCL/ExtSeqTrack.cpp:56]: yeah let's go modern

File: avr/cores/megacommand/CommonTools/NibbleArray.h (hunk 2 above)
Comment (yatli) [reply avr/cores/megacommand/CommonTools/NibbleArray.h:48]: Nope, also copied.

File: avr/cores/megacommand/CommonTools/NibbleArray.h (hunk 1 above)
Comment (jmamma) [reply avr/cores/megacommand/CommonTools/NibbleArray.h:54]: (jmamma) c+ -> (yatli) c++

Comment (jmamma) [reply avr/cores/megacommand/CommonTools/NibbleArray.h:54]: (jmamma) c+ -> (yatli) c++

**Comment by jmamma** [comment]
Quick pass 
70d75c6

Just tried recording the A4 from chromatic page and doesn't sound like anything is being recorded.

Need to refuel. 🍔

**Comment by jmamma** [comment]
Crash:

Chromatic page.. record a few notes then press shift1 (page_select)

**Comment by jmamma** [comment]
```
Sending note
pitch = 21
note_idx = 65535
note_num = 45
step_count = 1
"here" = here
note_idx = 37896
```

Bad idx ?


```
340 
341   if (note_idx == 0xFFFF) {
342     DEBUG_DUMP(note_idx);
343     DEBUG_DUMP(note_num);
344     DEBUG_DUMP(step_count);
345  
346           note_idx = add_event(step_count);
347   } 
348   DEBUG_DUMP("here");
349   DEBUG_DUMP(note_idx);
350   if (note_idx != 0xFFFF) {
351    events[note_idx].is_lock = false;
352     events[note_idx].event_on = true;
353     events[note_idx].event_value = note_num;
354     events[note_idx].cond_id = condition;
355     events[note_idx].micro_timing = utiming;
356   }
357 }
```

**Comment by jmamma** [comment]
looks like missing return.

**Comment by jmamma** [comment]
Got some note ouput now. but something still not right.

**Comment by jmamma** [comment]
Handing back over to you 💤

File: avr/cores/megacommand/MCL/ExtSeqTrack.cpp [hunk 4]
```diff
@@ -31,6 +29,69 @@ void ExtSeqTrack::re_sync() {
...
+  uint8_t u = timing_buckets.get(step);
+  if (15 == u || event_count >= NUM_EXT_EVENTS) {
+    // bucket full or track full
+    return 0xFFFF;
+  }
+  ++event_count;
+  timing_buckets.set(step, u + 1);
+  uint16_t idx, end;
+  locate(step, idx, end);
+  memmove(events + end + 1, events + end,
```
Comment (jmamma) [inline avr/cores/megacommand/MCL/ExtSeqTrack.cpp:57]: Is this shifting correct?

I thought the new event is being placed at events + idx?

Comment (yatli) [reply avr/cores/megacommand/MCL/ExtSeqTrack.cpp:57]: should be good. I'm inserting at the end of the current bucket.

Comment (yatli) [reply avr/cores/megacommand/MCL/ExtSeqTrack.cpp:57]: The memmove size can be reduced to sizeof(ext_event_t) * (event_count - end - 1) though

File: avr/cores/megacommand/MCL/ExtSeqTrack.cpp [hunk 5]
```diff
@@ -44,27 +105,38 @@ void ExtSeqTrack::seq() {
...
+    uint16_t ev_idx, ev_end;
+
+    // Locate CURRENT
+    locate(step_count, ev_idx, ev_end);
+
+    // Go over CURRENT
+    for (; ev_idx != ev_end; ++ev_idx) {
+      auto u = events[ev_idx].micro_timing;
+      if (u >= timing_mid && u - timing_mid == mod12_counter) {
+        handle_event(ev_idx);
```
Comment (jmamma) [inline avr/cores/megacommand/MCL/ExtSeqTrack.cpp:123]: fyi: when lock events are added, they need to be sent before note_on/off .
This ensures machine params are updated before sound trigger.

Comment (yatli) [reply avr/cores/megacommand/MCL/ExtSeqTrack.cpp:123]: ah, yes.. need to do it in the editor part

Comment (yatli) [reply avr/cores/megacommand/MCL/ExtSeqTrack.cpp:123]: also, velocity locks should be treated specially because that's part of the midi note event.

**Comment by jmamma** [comment]
Just doing some more testing.

If I record an arbitrarily long sequence greater than 1 note, the first note is never played back.

**Comment by jmamma** [comment]
A few things. timing_buckets is not initialized on creation? 

```
121   void locate(uint8_t step, uint16_t& ev_idx, uint16_t& ev_end) {
122     ev_idx = 0;
123     ev_end = timing_buckets.get(step_count);
124     for (uint8_t i = 0; i < step_count; ++i) {
125       ev_idx += timing_buckets.get(i);
126     }
127 
128     ev_end += ev_idx;
130   }
```

Edit: ev_end is max range, this value is never used as index.

--

On SeqExtStepPage the note detection routine only seems to run when the ISR triggers note out messages on port 2. This is extremely weird. The note detection is not working correctly, but it flashing bogus notes when note_on/off is transmitted.

File: avr/cores/megacommand/MCL/ExtSeqTrack.cpp [hunk 6]
```diff
@@ -31,6 +29,69 @@ void ExtSeqTrack::re_sync() {
...
+  // step is at least 1, move back to get current step
+  --step;
+  timing_buckets.set(step, bucket - 1);
+  memmove(events + index, events + index + 1,
+          sizeof(ext_event_t) * (NUM_EXT_EVENTS - index - 1));
+  --event_count;
+}
+
+uint16_t ExtSeqTrack::add_event(uint8_t step) {
+  uint8_t u = timing_buckets.get(step);
```
Comment (jmamma) [inline avr/cores/megacommand/MCL/ExtSeqTrack.cpp:49]: shouldn't this line, be placed after the memmove.

The locate(step,idx,end) command will return u + 1

File: avr/cores/megacommand/MCL/ExtSeqTrack.cpp [hunk 7]
```diff
@@ -31,6 +29,69 @@ void ExtSeqTrack::re_sync() {
...
+  if (15 == u || event_count >= NUM_EXT_EVENTS) {
+    // bucket full or track full
+    return 0xFFFF;
+  }
+  ++event_count;
+  timing_buckets.set(step, u + 1);
```
Comment (jmamma) [inline avr/cores/megacommand/MCL/ExtSeqTrack.cpp]: shouldn't this line, be placed after the memmove.

The locate(step,idx,end) command will set end = u + 1

File: avr/cores/megacommand/MCL/ExtSeqTrack.h [hunk 8]
```diff
@@ -120,11 +120,13 @@ class ExtSeqTrack : public ExtSeqTrackData, public SeqTrack {
 
   void locate(uint8_t step, uint16_t& ev_idx, uint16_t& ev_end) {
     ev_idx = 0;
-    ev_end = timing_buckets.get(step_count);
-    for (uint8_t i = 0; i < step_count; ++i) {
+    ev_end = timing_buckets.get(step);
+    for (uint8_t i = 0; i < step; ++i) {
```
Comment (jmamma) [inline avr/cores/megacommand/MCL/ExtSeqTrack.h:123]: 😱 @yatli

**Comment by jmamma** [comment]
Ext live record is now working 🚀 🌔 

ExtSeqStepPage is showing the recorded notes 👍

**Pull request merged**